File: src/main/java/com/turleylabs/algo/trader/kata/RefactorMeAlgorithm.java
package com.turleylabs.algo.trader.kata;

import com.turleylabs.algo.trader.kata.framework.*;

import java.time.LocalDate;
import java.util.Map;

public class RefactorMeAlgorithm extends BaseAlgorithm {

    String symbol = "TQQQ";
    SimpleMovingAverage movingAverage200;
    SimpleMovingAverage movingAverage50;
    SimpleMovingAverage movingAverage21;
    SimpleMovingAverage movingAverage10;
    double previousMovingAverage50;
    double previousMovingAverage21;
    double previousMovingAverage10;
    double previousPrice;
    LocalDate previous;
    CBOE lastVix;
    boolean boughtBelow50;
    boolean tookProfits;

    public void initialize() {
        this.setStartDate(2010, 3, 23);  //Set Start Date
        this.setEndDate(2020, 03, 06);

        this.setCash(100000);             //Set Strategy Cash

        movingAverage200 = this.SMA(symbol, 200);
        movingAverage50 = this.SMA(symbol, 50);
        movingAverage21 = this.SMA(symbol, 21);
        movingAverage10 = this.SMA(symbol, 10);

    }

    protected void onData(Slice data) {
        if (data.getCBOE("VIX") != null) {
            lastVix = data.getCBOE("VIX");
        }
        if (previous == getDate()) return;

        if (!movingAverage200.isReady()) return;

        if (data.get(symbol) == null) {
            return;
        }
        if (tookProfits) {
            if (data.get(symbol).getPrice() < movingAverage10.getValue()) {
                tookProfits = false;
            }
        } else if (portfolio.getOrDefault(symbol, Holding.Default).getQuantity() == 0) {

            if (data.get(symbol).getPrice() > movingAverage10.getValue()
                    && movingAverage10.getValue() > movingAverage21.getValue()
                    && movingAverage10.getValue() > previousMovingAverage10
                    && movingAverage21.getValue() > previousMovingAverage21
                    && (double) (lastVix.getClose()) < 19.0
                    && !(data.get(symbol).getPrice() >= (movingAverage50.getValue() * 1.15) && data.get(symbol).getPrice() >= (movingAverage200.getValue() * 1.40))
                    && (data.get(symbol).getPrice() - movingAverage10.getValue()) / movingAverage10.getValue() < 0.07) {
                this.log(String.format("Buy %s Vix %.4f. above 10 MA %.4f", symbol, lastVix.getClose(), (data.get(symbol).getPrice() - movingAverage10.getValue()) / movingAverage10.getValue()));
                double amount = 1.0;
                this.setHoldings(symbol, amount);

                boughtBelow50 = data.get(symbol).getPrice() < movingAverage50.getValue();
            }
        } else {
            double change = (data.get(symbol).getPrice() - portfolio.get(symbol).getAveragePrice()) / portfolio.get(symbol).getAveragePrice();

            if (data.get(symbol).getPrice() < (movingAverage50.getValue() * .93) && !boughtBelow50) {
                this.log(String.format("Sell %s loss of 50 day. Gain %.4f. Vix %.4f", symbol, change, lastVix.getClose()));
                this.liquidate(symbol);
            } else {
                if ((double) (lastVix.getClose()) > 22.0) {
                    this.log(String.format("Sell %s high volatility. Gain %.4f. Vix %.4f", symbol, change, lastVix.getClose()));
                    this.liquidate(symbol);
                } else {
                    if (movingAverage10.getValue() < 0.97 * movingAverage21.getValue()) {
                        this.log(String.format("Sell %s 10 day below 21 day. Gain %.4f. Vix %.4f", symbol, change, lastVix.getClose()));
                        this.liquidate(symbol);
                    } else {
                        if (data.get(symbol).getPrice() >= (movingAverage50.getValue() * 1.15) && data.get(symbol).getPrice() >= (movingAverage200.getValue() * 1.40)) {
                            this.log(String.format("Sell %s taking profits. Gain %.4f. Vix %.4f", symbol, change, lastVix.getClose()));
                            this.liquidate(symbol);
                            tookProfits = true;
                        }
                    }
                }
            }
        }

        previous = getDate();
        previousMovingAverage50 = movingAverage50.getValue();
        previousMovingAverage21 = movingAverage21.getValue();
        previousMovingAverage10 = movingAverage10.getValue();
        previousPrice = data.get(symbol).getPrice();
    }

    @Override
    public String toString() {
        String approveString = portfolioToString();
        approveString += tradesToString();
        return approveString;
    }

    private String tradesToString() {
        return "Trades{\n" +
                trades.stream().map(this::tradeToString).reduce("", (left, right) -> left + right + "\n") +
                "}\n";
    }

    private String tradeToString(Trade trade) {
        return "[Date=" + trade.getDate() + ", Ticker=" + trade.getSymbol() + ", Quantity=" + trade.getNumberOfShares() + ", Price=" + trade.getAveragePrice() + "]";
    }

    public String portfolioToString() {
        return "Portfolio{\n" +
                portfolio.entrySet().stream().map(this::holdingToString).reduce("", (left, right) -> left + right + "\n") +
                "}\n";
    }

    private String holdingToString(Map.Entry<String, Holding> entry) {
        return "[Ticker=" + entry.getKey() + ", Quantity=" + entry.getValue().getQuantity() + ", Price=" +  entry.getValue().getAveragePrice() + "]";
    }

}

File: typescript/src/refactor_me_algorithm.ts
import { BaseAlgorithm } from './framework/base_algorithm';
import { SimpleMovingAverage } from './framework/simple_moving_average';
import { CBOE } from './framework/cboe';
import { Slice } from './framework/slice';
import { Holding } from './framework/holding';

function round4(n: number): string { 
  return n.toFixed(4);
}

export class RefactorMeAlgorithm extends BaseAlgorithm {
  symbol: string = 'TQQQ';
  movingAverage200: SimpleMovingAverage;
  movingAverage50: SimpleMovingAverage;
  movingAverage21: SimpleMovingAverage;
  movingAverage10: SimpleMovingAverage;
  previousMovingAverage50: number = 0;
  previousMovingAverage21: number = 0;
  previousMovingAverage10: number = 0;
  previousPrice: number = 0;
  previous: Date;
  lastVix: CBOE;
  boughtBelow50: boolean = false;
  tookProfits: boolean = false;

  initialize() {
    this.setStartDate(2010, 3, 23);
    this.setEndDate(2020, 3, 6);

    this.setCash(100000);

    this.movingAverage200 = this.SMA(this.symbol, 200);
    this.movingAverage50 = this.SMA(this.symbol, 50);
    this.movingAverage21 = this.SMA(this.symbol, 21);
    this.movingAverage10 = this.SMA(this.symbol, 10);
  }

  protected onData(data: Slice) {
    if (data.getCBOE('VIX') != null) {
      this.lastVix = data.getCBOE('VIX');
    }
    if (
      (this.previous && this.previous.toISOString().split('T')[0]) ===
      this.getDate().toISOString().split('T')[0]
    ) {
      return;
    }
    if (!this.movingAverage200.isReady) {
      return;
    }

    if (data.get(this.symbol) == null) {
      return;
    }

    if (this.tookProfits) {
      if (data.get(this.symbol).price < this.movingAverage10.value) {
        this.tookProfits = false;
      }
    } else if ((this.portfolio[this.symbol] || Holding.Default).quantity === 0) {
      if (
        data.get(this.symbol).price > this.movingAverage10.value &&
        this.movingAverage10.value > this.movingAverage21.value &&
        this.movingAverage10.value > this.previousMovingAverage10 &&
        this.movingAverage21.value > this.previousMovingAverage21 &&
        this.lastVix.close < 19.0 &&
        !(data.get(this.symbol).price >= (this.movingAverage50.value * 1.15) && data.get(this.symbol).price >= (this.movingAverage200.value * 1.40)) &&
        (data.get(this.symbol).price - this.movingAverage10.value) / this.movingAverage10.value < 0.07
      ) {
        this.log(`Buy ${this.symbol} Vix ${round4(this.lastVix.close)}. above 10 MA ${round4((data.get(this.symbol).price - this.movingAverage10.value) / this.movingAverage10.value)}`);
        const amount = 1.0;
        this.setHoldings(this.symbol, amount);
        this.boughtBelow50 = data.get(this.symbol).price < this.movingAverage50.value;
      }
    } else {
      const change = (data.get(this.symbol).price - this.portfolio[this.symbol].averagePrice) / this.portfolio[this.symbol].averagePrice;

      if (data.get(this.symbol).price < (this.movingAverage50.value * 0.93) && !this.boughtBelow50) {
        this.log(`Sell ${this.symbol} loss of 50 day. Gain ${round4(change)}. Vix ${round4(this.lastVix.close)}`);
        this.liquidate(this.symbol);
      } else {
        if (this.lastVix.close > 22.0) {
          this.log(`Sell ${this.symbol} high volatility. Gain ${round4(change)}. Vix ${round4(this.lastVix.close)}`);
          this.liquidate(this.symbol);
        } else {
          if (this.movingAverage10.value < 0.97 * this.movingAverage21.value) {
            this.log(`Sell ${this.symbol} 10 day below 21 day. Gain ${round4(change)}. Vix ${round4(this.lastVix.close)}`);
            this.liquidate(this.symbol);
          } else {
            if (data.get(this.symbol).price >= (this.movingAverage50.value * 1.15) && data.get(this.symbol).price >= (this.movingAverage200.value * 1.40)) {
              this.log(`Sell ${this.symbol} taking profits. Gain ${round4(change)}. Vix ${round4(this.lastVix.close)}`);
              this.liquidate(this.symbol);
              this.tookProfits = true;
            }
          }
        }
      }
    }

    this.previous = this.getDate();
    this.previousMovingAverage50 = this.movingAverage50.value;
    this.previousMovingAverage21 = this.movingAverage21.value;
    this.previousMovingAverage10 = this.movingAverage10.value;
    this.previousPrice = data.get(this.symbol).price;
  }
}